sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatRecordReaderFn.java

```diff
@@ -0,0 +1,204 @@
+import org.apache.hive.hcatalog.data.HCatRecord;
+import org.apache.hive.hcatalog.data.transfer.DataTransferFactory;
+import org.apache.hive.hcatalog.data.transfer.HCatReader;
+import org.apache.hive.hcatalog.data.transfer.ReadEntity;
+import org.apache.hive.hcatalog.data.transfer.ReaderContext;
+import org.joda.time.Instant;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Unbounded reader for doing reads from Hcat. */
+@BoundedPerElement
+class HCatRecordReaderFn extends DoFn<HCatRecordReaderFn.PartitionWrapper, HCatRecord> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HCatRecordReaderFn.class);
+
+  /** Wrapper class for partition data. */
+  public static class PartitionWrapper implements Serializable {
+    final Partition partition;
+    final HCatalogIO.Read readSpec;
+    final String watermarkPartitionColumn;
+    final ImmutableList<String> partitionCols;
+    final SerializableFunction<String, Instant> watermarkerFn;
+
+    public PartitionWrapper(
+        Partition partition,
+        HCatalogIO.Read readSpec,
+        String watermarkPartitionColumn,
+        ImmutableList<String> partitionCols,
+        SerializableFunction<String, Instant> watermarkerFn) {
+
+      this.partition = partition;
+      this.readSpec = readSpec;
+      this.watermarkPartitionColumn = watermarkPartitionColumn;
+      this.partitionCols = partitionCols;
+      this.watermarkerFn = watermarkerFn;
+    }
+  }
+
+  private ReaderContext getReaderContext(PartitionWrapper wrapper, long desiredSplitCount)
+      throws HCatException {
+    final HCatalogIO.Read spec = wrapper.readSpec;
+    final Partition partition = wrapper.partition;
+    final List<String> values = partition.getValues();
+    final ImmutableList<String> partitionCols = wrapper.partitionCols;
+    checkArgument(
+        values.size() == partitionCols.size(),
+        "Number of input partitions should be equal to the values of list partition values.");
+
+    List<String> filter = new ArrayList<>();
+    for (int i = 0; i < partitionCols.size(); i++) {
+      filter.add(partitionCols.get(i) + "=" + "'" + values.get(i) + "'");
+    }
+    final String filterString = String.join(" and ", filter);
+
+    ReadEntity entity =
+        new ReadEntity.Builder()
+            .withDatabase(spec.getDatabase())
+            .withFilter(filterString)
+            .withTable(spec.getTable())
+            .build();
+    // pass the 'desired' split count as an hint to the API
+    Map<String, String> configProps = new HashMap<>(spec.getConfigProperties());
+    configProps.put(
+        HCatConstants.HCAT_DESIRED_PARTITION_NUM_SPLITS, String.valueOf(desiredSplitCount));
+    return DataTransferFactory.getHCatReader(entity, configProps).prepareRead();
+  }
+
+  private long getFileSizeForPartitions(PartitionWrapper partitionWrapper) throws Exception {
+    IMetaStoreClient client = null;
+    Configuration conf = new Configuration();
```

Comment:
Can you extract the code that creates a new HiveMetastoreClient into a new static method in a class called `HCatalogUtils` and use it here and in `PartitionPollerFn` too. Extra points if you change it in `HCatalogIO.getEstimatedSizeBytes` too.